perun-registrar-lib/src/main/java/cz/metacentrum/perun/registrar/impl/MailManagerImpl.java

```diff
@@ -2529,45 +2516,90 @@ private void setHtmlMessageWithAltPlainTextMessage (MimeMessage message, String
 	}
 
 	/**
-	 * Get notification emails for given VO. If toEmail is not specified,
+	 * Get notification emails for given VO. If toEmail (URN_VO_TO_EMAIL) has no value set,
 	 * all configured roles for receiving VO notifications are used.
 	 *
-	 * @param app  Application to get VO from
-	 * @param voEmailAttr  String name of given email attribute
-	 * @return list of emails used for sending notifications.
+	 * @param vo Virtual Organization for which the lookup is performed. If null, empty result will be returned.
+	 * @return List of emails used for sending notifications (can be empty).
 	 */
-	private List<String> getMailsFromVo(Application app, String voEmailAttr)
-		throws AttributeNotExistsException, WrongAttributeAssignmentException {
-
-		List<String> emails = new ArrayList<>();
+	private Set<String> getToEmailsVoLevel(Vo vo)
+		throws AttributeNotExistsException, WrongAttributeAssignmentException
+	{
+		if (vo == null) {
+			return new HashSet<>();
+		}
+		Attribute toEmailAttr = attrManager.getAttribute(registrarSession, vo, URN_VO_TO_EMAIL);
+		Set<String> emails = getToEmailsFromAttribute(toEmailAttr);
+		if (emails.isEmpty()) {
+			emails = getToEmailsViaRoles(VO, vo);
+		}
+		return emails;
+	}
 
-		Attribute attrToEmail = attrManager.getAttribute(registrarSession, app.getVo(), voEmailAttr);
+	/**
+	 * Get notification emails for given Group. If toEmail (URN_GROUP_TO_EMAIL) has no value set,
+	 * all configured roles for receiving GROUP notifications are used.
+	 *
+	 * @param group Group for which the lookup is performed. If null, empty result will be returned.
+	 * @return List of emails used for sending notifications (can be empty).
+	 */
+	private Set<String> getToEmailsGroupLevel(Group group)
+		throws AttributeNotExistsException, WrongAttributeAssignmentException
+	{
+		if (group == null) {
+			return new HashSet<>();
+		}
+		Attribute toEmailAttr = attrManager.getAttribute(registrarSession, group, URN_GROUP_TO_EMAIL);
+		Set<String> emails = getToEmailsFromAttribute(toEmailAttr);
+		if (emails.isEmpty()) {
+			emails = getToEmailsViaRoles(GROUP, group);
+		}
+		return emails;
+	}
 
+	/**
+	 * Get "TO" email addresses of notification recipient(s) from the attribute
+	 * @param toEmailAttr Attribute with the recipients as value
+	 * @return Set of identified emails. Can be empty if attr has no non-blank item in array value or no value at all.
+	 */
+	private Set<String> getToEmailsFromAttribute(Attribute toEmailAttr) {
 		// if there are any pre-defined VO emailTo attributes, use them
-		if (attrToEmail != null && attrToEmail.getValue() != null) {
-			ArrayList<String> value = attrToEmail.valueAsList();
+		Set<String> emails = new HashSet<>();
+		if (toEmailAttr != null && toEmailAttr.getValue() != null) {
+			ArrayList<String> value = toEmailAttr.valueAsList();
 			for (String adr : value) {
-				if (adr != null && !adr.isEmpty()) {
+				if (StringUtils.hasText(adr)) {
```

Comment:
The attribute module should probably do this check. It is this way for VO toEmail/fromEmail attrs, toEmail group attribute module is missing for some reason.
You can consider adding it but I'm okay with leaving it as it is (since it's registrar)